Show Largest Contentful Paints in the timeline and event list

smfr opened the pull request:
#### 6e6554f10c46eaa882c721c1c22cbdbfb1c72d17
<pre>
Show Largest Contentful Paints in the timeline and event list
<a href="https://bugs.webkit.org/show_bug.cgi?id=301450">https://bugs.webkit.org/show_bug.cgi?id=301450</a>
<a href="https://rdar.apple.com/160480088">rdar://160480088</a>

Reviewed by Devin Rousso.

Add support for displaying &quot;First Contentful Paint&quot; and &quot;Largest Contentful Paint&quot; in the
web inspector timeline.

Add `InspectorInstrumentation::didEnqueue...()` which is called when we enqueue one of
these paint entries. The LCP entry passes the element and an area, so inspector can show
these.

`InspectorTimelineAgent::didEnqueue...()` then creates timeline entries with the right
type, adding the metadata as a appropriate.

To `inspector/protocol/Timeline.json` we add two new event types. When `TimelineManager`
encounters one of these, it creates a `WI.LayoutTimelineRecord` for FCP or LCP. The latter
takes the extra metadata.

`LayoutTimelineRecord` gains the ability to store a `domNode` and `area`. Area cannot be
derived from a quad here, because the LCP area results in math on a scalar value[1], so we
have to store area separately.

Enhance `LayoutTimelineDataGridNode` to be able to create a &quot;go to arrow&quot; which provides a
button to jump to the element when hovered.

I chose a teal color with a `P` for the performance entry icons, and used teal as the
timeline pip color.

[1] <a href="https://w3c.github.io/largest-contentful-paint/#sec-effective-visual-size">https://w3c.github.io/largest-contentful-paint/#sec-effective-visual-size</a>

* Source/JavaScriptCore/inspector/protocol/Timeline.json:
* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::didEnqueueFirstContentfulPaintImpl):
(WebCore::InspectorInstrumentation::didEnqueueLargestContentfulPaintImpl):
* Source/WebCore/inspector/InspectorInstrumentation.h:
(WebCore::InspectorInstrumentation::didEnqueueFirstContentfulPaint):
(WebCore::InspectorInstrumentation::didEnqueueLargestContentfulPaint):
* Source/WebCore/inspector/TimelineRecordFactory.cpp:
(WebCore::TimelineRecordFactory::createLargestContentfulPaintData):
* Source/WebCore/inspector/TimelineRecordFactory.h:
* Source/WebCore/inspector/agents/InspectorTimelineAgent.cpp:
(WebCore::InspectorTimelineAgent::didEnqueueFirstContentfulPaint):
(WebCore::InspectorTimelineAgent::didEnqueueLargestContentfulPaint):
(WebCore::toProtocol):
* Source/WebCore/inspector/agents/InspectorTimelineAgent.h:
* Source/WebCore/page/Performance.cpp:
(WebCore::Performance::reportFirstContentfulPaint):
(WebCore::Performance::enqueueLargestContentfulPaint):
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Controllers/TimelineManager.js:
(WI.TimelineManager.prototype._processRecord):
* Source/WebInspectorUI/UserInterface/Images/TypeIcons.svg:
* Source/WebInspectorUI/UserInterface/Models/LayoutTimelineRecord.js:
(WI.LayoutTimelineRecord.displayNameForEventType):
(WI.LayoutTimelineRecord.async fromJSON):
(WI.LayoutTimelineRecord.prototype.get area):
(WI.LayoutTimelineRecord.prototype.get domNode):
* Source/WebInspectorUI/UserInterface/Views/CPUTimelineView.js:
(WI.CPUTimelineView.prototype._attemptSelectIndicatatorTimelineRecord):
* Source/WebInspectorUI/UserInterface/Views/LayoutTimelineDataGridNode.js:
(WI.LayoutTimelineDataGridNode.prototype.get data):
(WI.LayoutTimelineDataGridNode.prototype.createCellContent):
(WI.LayoutTimelineDataGridNode):
* Source/WebInspectorUI/UserInterface/Views/LayoutTimelineView.js:
(WI.LayoutTimelineView.prototype._showHighlightForRecord):
* Source/WebInspectorUI/UserInterface/Views/TimelineIcons.css:
(.performance-entry .icon):
(@media (prefers-color-scheme: dark) .performance-entry .icon):
* Source/WebInspectorUI/UserInterface/Views/TimelineRecordBar.css:
(.timeline-record-bar.timeline-record-type-layout:is(.first-contentful-paint, .largest-contentful-paint) &gt; .segment):
* Source/WebInspectorUI/UserInterface/Views/TimelineRecordTreeElement.js:
* Source/WebInspectorUI/UserInterface/Views/TimelineTabContentView.js:
(WI.TimelineTabContentView.iconClassNameForRecord):

Canonical link: <a href="https://commits.webkit.org/302248@main">https://commits.webkit.org/302248@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b934b469c4aa99d02f776c6c4939df520da0064e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128322 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/598 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39154 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135713 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79791 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/c67d847f-1638-4d4d-8019-6f1fdd876f69) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130194 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/534 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/472 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97675 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65580 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/f16b590f-aef3-4048-b4c8-b1ff1a346de1) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131270 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/361 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114958 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78266 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/06fe16fe-35b7-4703-899c-27bcea1d2ef5) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/344 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33065 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79001 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/120343 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108728 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33549 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138169 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/126782 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/443 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/419 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106213 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/482 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111299 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106013 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27046 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/360 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29849 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/52755 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/494 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/63666 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/159806 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/393 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39916 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/460 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/456 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-commit-queue commented:
Commit message contains (OOPS!) and no valid reviewer found, blocking PR #52976. Details: [Build #26962](https://ews-build.webkit.org/#/builders/19/builds/26962)

dcrousso commented on Source/WebCore/page/Performance.cpp:
```diff
@@ -375,13 +376,19 @@ void Performance::setResourceTimingBufferSize(unsigned size)
 void Performance::enqueueLargestContentfulPaint(Ref<LargestContentfulPaint>&& paintEntry)
 {
+    if (RefPtr context = scriptExecutionContext())
+        InspectorInstrumentation::didEnqueuePerformanceEntry(*context, "largest-contentful-paint"_s, paintEntry->element(), paintEntry->size());
```
NIT: in order to avoid doing extra work when Web Inspector is not open, id suggest just passing either the entire `Performance` object or the `ScriptExecutionContext*`
```cpp
InspectorInstrumentation::enqueueLargestContentfulPaint(*this, paintEntry);
```
this way, we can delay the work of unpacking anything until we actually get to the `InspectorTimelineAgent`, which will only happen if Web Inspector is open (see `FAST_RETURN_IF_NO_FRONTENDS`)

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/LayoutTimelineRecord.js:
```diff
@@ -37,6 +37,8 @@ WI.LayoutTimelineRecord = class LayoutTimelineRecord extends WI.TimelineRecord
 
         this._eventType = eventType;
         this._quad = quad || null;
+        this._area = area || null;
```
is there really no way of getting a `quad` for LCP?  without this it wont be possible to highlight the layout area when the developer hovers over the record in the Layout timeline view :(

since we are providing a `domNode` we could have it at least highlight that instead?  id add something like this to `WI.LayoutTimelineView.prototype._showHighlightForRecord`
```js

```

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/LayoutTimelineRecord.js:
```diff
@@ -103,14 +109,22 @@ WI.LayoutTimelineRecord = class LayoutTimelineRecord extends WI.TimelineRecord
+        if (this._quad)
+            return this.width * this.height;
+
+        return this._area ? this._area : NaN;
```
Style: this can be simplified using nullish coalescing
```js
return this._area ?? NaN;
```

dcrousso replied:
Style: this can be simplified using nullish coalescing
```js
return this._area ?? (this.width * this.height);
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/LayoutTimelineDataGridNode.js:
```diff
@@ -60,6 +60,19 @@ WI.LayoutTimelineDataGridNode = class LayoutTimelineDataGridNode extends WI.Time
+
+                if (this.record.domNode) {
+                    let goToButton = WI.createGoToArrowButton();
+                    fragment.appendChild(goToButton);
```
NIT: you can combine these lines
```js
let goToArrow = fragment.appendChild(WI.createGoToArrowButton());
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/LayoutTimelineDataGridNode.js:
```diff
@@ -60,6 +60,19 @@ WI.LayoutTimelineDataGridNode = class LayoutTimelineDataGridNode extends WI.Time
+                if (this.record.domNode) {
+                    let goToButton = WI.createGoToArrowButton();
+                    fragment.appendChild(goToButton);
+                    goToButton.addEventListener("click", this._goToArrowWasClicked.bind(this), false);
```
the `false` is not necessary

also, rather than use `bind` we could just use an arrow function
```js
goToArrow.addEventListener("click", (event) => {
    WI.showMainFrameDOMTree(domNode, {ignoreSearchTab: true});
});
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/LayoutTimelineDataGridNode.js:
```diff
@@ -81,4 +94,11 @@ WI.LayoutTimelineDataGridNode = class LayoutTimelineDataGridNode extends WI.Time
+
+    _goToArrowWasClicked()
+    {
+        var domNode = WI.domManager.nodeForId(this.record.domNode);
```
Style: please use `let`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/TimelineRecordBar.css:
```diff
@@ -139,6 +139,12 @@ body[dir=rtl] .timeline-record-bar > .segment:last-of-type {
 }
 
+.timeline-record-bar.timeline-record-type-layout.first-contentful-paint > .segment,
+.timeline-record-bar.timeline-record-type-layout.largest-contentful-paint > .segment {
```
Style: we can use `:is`
```css
.timeline-record-bar.timeline-record-type-layout:is(.first-contentful-paint, .largest-contentful-paint) > .segment {
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/LayoutTimelineRecord.js:
```diff
@@ -126,7 +140,9 @@ WI.LayoutTimelineRecord.EventType = {
-    Composite: "composite"
+    Composite: "composite",
+    FirstContentfulPaint: "first-contentful-paint",
+    LargestContentfulPaint: "largest-contentful-paint"
```
Style: please add a trailing `,`

smfr committed:
Show Largest Contentful Paints in the timeline and event list
https://bugs.webkit.org/show_bug.cgi?id=301450
rdar://160480088

Reviewed by Devin Rousso.

Add support for displaying "First Contentful Paint" and "Largest Contentful Paint" in the
web inspector timeline.

Add `InspectorInstrumentation::didEnqueue...()` which is called when we enqueue one of
these paint entries. The LCP entry passes the element and an area, so inspector can show
these.

`InspectorTimelineAgent::didEnqueue...()` then creates timeline entries with the right
type, adding the metadata as a appropriate.

To `inspector/protocol/Timeline.json` we add two new event types. When `TimelineManager`
encounters one of these, it creates a `WI.LayoutTimelineRecord` for FCP or LCP. The latter
takes the extra metadata.

`LayoutTimelineRecord` gains the ability to store a `domNode` and `area`. Area cannot be
derived from a quad here, because the LCP area results in math on a scalar value[1], so we
have to store area separately.

Enhance `LayoutTimelineDataGridNode` to be able to create a "go to arrow" which provides a
button to jump to the element when hovered.

I chose a teal color with a `P` for the performance entry icons, and used teal as the
timeline pip color.

[1] https://w3c.github.io/largest-contentful-paint/#sec-effective-visual-size

* Source/JavaScriptCore/inspector/protocol/Timeline.json:
* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::didEnqueueFirstContentfulPaintImpl):
(WebCore::InspectorInstrumentation::didEnqueueLargestContentfulPaintImpl):
* Source/WebCore/inspector/InspectorInstrumentation.h:
(WebCore::InspectorInstrumentation::didEnqueueFirstContentfulPaint):
(WebCore::InspectorInstrumentation::didEnqueueLargestContentfulPaint):
* Source/WebCore/inspector/TimelineRecordFactory.cpp:
(WebCore::TimelineRecordFactory::createLargestContentfulPaintData):
* Source/WebCore/inspector/TimelineRecordFactory.h:
* Source/WebCore/inspector/agents/InspectorTimelineAgent.cpp:
(WebCore::InspectorTimelineAgent::didEnqueueFirstContentfulPaint):
(WebCore::InspectorTimelineAgent::didEnqueueLargestContentfulPaint):
(WebCore::toProtocol):
* Source/WebCore/inspector/agents/InspectorTimelineAgent.h:
* Source/WebCore/page/Performance.cpp:
(WebCore::Performance::reportFirstContentfulPaint):
(WebCore::Performance::enqueueLargestContentfulPaint):
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Controllers/TimelineManager.js:
(WI.TimelineManager.prototype._processRecord):
* Source/WebInspectorUI/UserInterface/Images/TypeIcons.svg:
* Source/WebInspectorUI/UserInterface/Models/LayoutTimelineRecord.js:
(WI.LayoutTimelineRecord.displayNameForEventType):
(WI.LayoutTimelineRecord.async fromJSON):
(WI.LayoutTimelineRecord.prototype.get area):
(WI.LayoutTimelineRecord.prototype.get domNode):
* Source/WebInspectorUI/UserInterface/Views/CPUTimelineView.js:
(WI.CPUTimelineView.prototype._attemptSelectIndicatatorTimelineRecord):
* Source/WebInspectorUI/UserInterface/Views/LayoutTimelineDataGridNode.js:
(WI.LayoutTimelineDataGridNode.prototype.get data):
(WI.LayoutTimelineDataGridNode.prototype.createCellContent):
(WI.LayoutTimelineDataGridNode):
* Source/WebInspectorUI/UserInterface/Views/LayoutTimelineView.js:
(WI.LayoutTimelineView.prototype._showHighlightForRecord):
* Source/WebInspectorUI/UserInterface/Views/TimelineIcons.css:
(.performance-entry .icon):
(@media (prefers-color-scheme: dark) .performance-entry .icon):
* Source/WebInspectorUI/UserInterface/Views/TimelineRecordBar.css:
(.timeline-record-bar.timeline-record-type-layout:is(.first-contentful-paint, .largest-contentful-paint) > .segment):
* Source/WebInspectorUI/UserInterface/Views/TimelineRecordTreeElement.js:
* Source/WebInspectorUI/UserInterface/Views/TimelineTabContentView.js:
(WI.TimelineTabContentView.iconClassNameForRecord):

Canonical link: https://commits.webkit.org/302248@main

webkit-commit-queue commented:
Committed 302248@main (6e6554f10c46): <https://commits.webkit.org/302248@main>

Reviewed commits have been landed. Closing PR #52976 and removing active labels.
